File: cpp/_200.cpp
// BFS
// O(MN) / O(MN)
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(), n = grid[0].size(), ans = 0;
        vector<vector<int>> dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (grid[r][c] == '1') {
                    ans++;
                    // BFS
                    queue<pair<int, int>> q;
                    q.push({r, c});
                    grid[r][c] = '0';
                    while (!q.empty()) {
                        pair<int, int> p = q.front();
                        q.pop();
                        for (int i = 0; i < 4; ++i) {
                            int nr = p.first + dirs[i][0], nc = p.second + dirs[i][1];
                            if (nr >= 0 && nr < m && nc >= 0 && nc < n && grid[nr][nc] == '1') {
                                grid[nr][nc] = '0';
                                q.push({nr, nc});
                            }
                        }
                    }
                }
            }
        }
        return ans;
    }
};


// DFS
// O(MN) / O(MN)
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(), n = grid[0].size(), ans = 0;
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (grid[r][c] == '1') {
                    dfs(grid, r, c);
                    ans++;
                }
            }
        }
        return ans;
    }
private:
    void dfs(vector<vector<char>>& grid, int r, int c) {
        int m = grid.size(), n = grid[0].size();
        if (r < 0 || r >= m || c < 0 || c >= n || grid[r][c] == '0')
            return;
        grid[r][c] = '0';
        dfs(grid, r - 1, c);
        dfs(grid, r + 1, c);
        dfs(grid, r, c - 1);
        dfs(grid, r, c + 1);
    }
};

File: python/_200.py
# Solution 1. BFS
# O(MN) / O(MN)
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        dirs = ((0, -1), (0, 1), (-1, 0), (1, 0))
        
        def bfs(r, c):
            queue = collections.deque([(r, c)])
            grid[r][c] = '0'
            while queue:
                r, c = queue.popleft()
                for dr, dc in dirs:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] == '1':
                        queue.append((nr, nc))
                        grid[nr][nc] = '0'
        
        ans = 0
        for r in range(m):
            for c in range(n):
                if grid[r][c] == '1':
                    bfs(r, c)
                    ans += 1
        return ans


# Solution 2. DFS - recursive
# O(MN) / O(MN)
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        dirs = ((0, -1), (0, 1), (-1, 0), (1, 0))
        
        def dfs(r, c):
            if r < 0 or r >= m or c < 0 or c >= n or grid[r][c] != '1':
                return
            grid[r][c] = '0'
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                dfs(nr, nc)
        
        ans = 0
        for r in range(m):
            for c in range(n):
                if grid[r][c] == '1':
                    dfs(r, c)
                    ans += 1
        return ans


# Solution 3. DFS - stack
# O(MN) / O(MN)
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        dirs = ((0, -1), (0, 1), (-1, 0), (1, 0))
            
        def dfs(r, c):
            stack = [(r, c)]
            grid[r][c] = '0'
            while stack:
                r, c = stack.pop()
                for dr, dc in dirs:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] == '1':
                        stack.append((nr, nc))
                        grid[nr][nc] = '0'

        ans = 0
        for r in range(m):
            for c in range(n):
                if grid[r][c] == '1':
                    dfs(r, c)
                    ans += 1
        return ans